src/test/java/TaskListServletTest.java

```diff
@@ -0,0 +1,137 @@
+  private static final TaskList validTaskList = new TaskList().setTitle(VALID_TASKLIST_TITLE);
+  private static final String VALID_TASKLIST_JSON = gson.toJson(validTaskList);
+
+  @Override
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+    tasksClientFactory = Mockito.mock(TasksClientFactory.class);
+    tasksClient = Mockito.mock(TasksClient.class);
+    servlet = new TaskListServlet(authenticationVerifier, tasksClientFactory);
+
+    Mockito.when(tasksClientFactory.getTasksClient(Mockito.any())).thenReturn(tasksClient);
+  }
+
+  @Test
+  public void getTasklists() throws Exception {
+    Mockito.when(tasksClient.listTaskLists()).thenReturn(SOME_TASKLISTS);
+    Mockito.when(tasksClient.listTasks(SOME_TASKLISTS.get(0))).thenReturn(TASKS_ONE_TWO);
+    Mockito.when(tasksClient.listTasks(SOME_TASKLISTS.get(1))).thenReturn(TASKS_THREE_FOUR);
+
+    HashMap<String, List<Task>> tasksWithTaskLists = new HashMap<>();
+    tasksWithTaskLists.put(SOME_TASKLISTS.get(0).getId(), TASKS_ONE_TWO);
+    tasksWithTaskLists.put(SOME_TASKLISTS.get(1).getId(), TASKS_THREE_FOUR);
+
+    JsonObject expectedResponseObject = new JsonObject();
+    expectedResponseObject.add("tasklists", gson.toJsonTree(SOME_TASKLISTS));
+    expectedResponseObject.add("tasks", gson.toJsonTree(tasksWithTaskLists));
+
+    String expectedResponse = gson.toJson(expectedResponseObject);
+
+    servlet.doGet(request, response);
+
+    Assert.assertTrue(stringWriter.toString().contains(expectedResponse));
+  }
+
+  @Test
+  public void getTasklistsEmpty() throws Exception {
+    Mockito.when(tasksClient.listTaskLists()).thenReturn(NO_TASKLISTS);
+
+    HashMap<String, List<Task>> emptyTasksWithTaskLists = new HashMap<>();
+
+    JsonObject expectedResponseObject = new JsonObject();
+    expectedResponseObject.add("tasklists", gson.toJsonTree(NO_TASKLISTS));
+    expectedResponseObject.add("tasks", gson.toJsonTree(emptyTasksWithTaskLists));
+
+    String expectedResponse = gson.toJson(expectedResponseObject);
+
+    servlet.doGet(request, response);
+
+    Assert.assertTrue(stringWriter.toString().contains(expectedResponse));
+  }
+
+  @Test
+  public void postTasklistNullNameGiven() throws Exception {
+    servlet.doPost(request, response);
+    Assert.assertEquals(400, response.getStatus());
+  }
+
+  @Test
+  public void postTasklistEmptyNameGiven() throws Exception {
+    Mockito.when(request.getParameter("taskListName")).thenReturn("");
+    servlet.doPost(request, response);
+    Assert.assertEquals(400, response.getStatus());
+  }
+
+  @Test
+  public void postTasklist() throws Exception {
+    Mockito.when(request.getParameter("taskListTitle")).thenReturn(VALID_TASKLIST_TITLE);
+    Mockito.when(tasksClient.postTaskList(VALID_TASKLIST_TITLE)).thenReturn(validTaskList);
+
+    servlet.doPost(request, response);
+    Assert.assertTrue(stringWriter.toString().contains(VALID_TASKLIST_JSON));
```

Comment:
I would recommend decoding the JSON and directly comparing it with `validTaskList`. Same for the other tests.

Reply:
Agreed - fixing now! Thanks!

Reply:
Update - I agree this would be ideal, but gson is giving me a ton of trouble when I try to do this. `TaskList` does not play well with gson when trying to deserialize JSON (Even when I do `TypeToken<TaskList>(){}.getType()`, I still get an error saying `java.lang.IllegalArgumentException: Map objects need to be parameterized unless you use a custom serializer. Use the com.google.gson.reflect.TypeToken to extract the ParameterizedType.`). 

I'm not sure how to proceed, and I want to check in with the team before working on this too much.

Reply:
Yes please do, there's other things you need to pull your hair over and I don't think this is one of them.